feat(baselines) Add Floco Baseline

dennis-grinwald opened the pull request:
## Issue

### Description

Add baseline Floco to Flower baselines. Reproduce SimpleCNN experiments on Five-Fold and Dirichlet CIFAR-10 data splits.

### Related issues/PRs

Link issues and/or PRs that are related to this PR.
Related issue [#4940](https://github.com/adap/flower/issues/4940#issue-2857741078)

## Proposal

### Explanation

### Checklist

- [ ] Implement proposed change
- [ ] Write tests
- [ ] Update [documentation](https://flower.ai/docs/writing-documentation.html)
- [ ] Make CI checks pass
- [ ] Ping maintainers on [Slack](https://flower.ai/join-slack/) (channel `#contributions`)

### Any other comments?

jafermarq reviewed (commented):
Hi @dennis-grinwald, this is a first pass through `Floco` which I was able to run following your `README.md` and I got the exact same results 🙌 

Below there are two main suggestion: (1) Could we make use of Flower Datasets to update the logic in your `dataset.py` or partially update it if creating a _fold_ partitioner for Flower Datasets isn't that straight forward; and (2) it would be beneficial to not require passing `num_clients` to the `ServerApp` as this makes some assumptions that might not hold when running Floco in a non-simulated setup.  More details below in my comments. The rest of the comments below are mostly suggestions for less important topics.

Let me know what you think!

dennis-grinwald commented:
Hi @jafermarq . Thank you for your detailed review and suggestions for improvement. I will have a look on it and get back to you with the changes soon.

dennis-grinwald commented on baselines/floco/README.md:
```diff
@@ -0,0 +1,111 @@
+
+**Datasets:** CIFAR-10
+
+**Hardware Setup:** We trained the models used for the plots below on a M1X chip.
```
Change this to Apple M1 Pro chip.

dennis-grinwald commented:
Hi @jafermarq. I've just committed some changes to this branch that address your mentioned improvement suggestions.

(1) I've implemented a FoldPartitioner and used the InnerDirichletPartitioner to comply with the Flower Datasets library. The updated Federated and Global test accuracy plots were generated using this code.
(2) Besides adapting the dataset logic to Flower Datasets, I've also changed the logic of loading the global testloader, as you have suggested.
(3) I've removed the server dependency on ```num_clients```.

I have also fixed your minor style suggestions. Let me know if anything else needs to be done.

jafermarq commented:
Hi @dennis-grinwald , many thanks for applying all those suggestions. The `Floco` PR looks like in a pretty good state. I was able to reproduce the results. Note i pushed some small changes (mostly cosmetic) to make some parts of the code more aligned with the latest version of Flower (related to the `Parameters` record creation in your `ClientApp`. All good!) and to solve some of the issues flagged by `pylint`.

I'll do a final pass later this week and merge it! 🚀

jafermarq commented on baselines/floco/README.md:
```diff
@@ -0,0 +1,112 @@
+---
+title: Federated Learning over Connected Modes
+url: https://openreview.net/pdf?id=JL2eMCfDW8
+labels: [personalization] 
```
@dennis-grinwald, what would be two other labels to use for `Floco`? I'm thinking of `cross-silo` and `simplex`

dennis-grinwald replied:
I'd say let's use cross-silo and linear mode connectivity :)

dennis-grinwald replied:
And thank you for putting in the work! Looking forward to seeing the method merged :)
If there is anything I could do to help, let me know!

jafermarq replied:
perfect, i added those labels

jafermarq commented on baselines/floco/README.md:
```diff
@@ -0,0 +1,112 @@
+---
+title: Federated Learning over Connected Modes
+url: https://openreview.net/pdf?id=JL2eMCfDW8
+labels: [personalization] 
```
```suggestion
labels: [personalization, cross-silo, linear mode connectivity] 
```

jafermarq commented on baselines/floco/README.md:
```diff
@@ -0,0 +1,112 @@
+
+**Task:** Image Classification
+
+**Model:** This repository implements a two-layer CNN as in \cite{FedAvg}. Additional models can be implemented by changing the `base_model` attribute in the `SimplexModel` class in `floco/model.py`.
```
```suggestion
**Model:** This repository implements a two-layer CNN as in [McMahan et al. 2017](https://arxiv.org/abs/1602.05629). Additional models can be implemented by changing the `base_model` attribute in the `SimplexModel` class in `floco/model.py`.
```

jafermarq replied:
this is what you meant by `\cite{FedAvg}` right?

dennis-grinwald replied:
Yes, correct. Thanks!

jafermarq commented on baselines/floco/README.md:
```diff
@@ -0,0 +1,112 @@
+
+**Hardware Setup:** We trained the models used for the plots below on an Apple M1 Pro chip.
+
+**Contributors:** <add here your names (or github tags, emails are fine too if you'd like)> 
```
I think this is the last remaining bit to add to the README.md.

dennis-grinwald replied:
Changed this.

jafermarq reviewed (commented):
Thanks @dennis-grinwald, just one minor point left to address below. Then i can merge `Floco` 🚀

jafermarq commented on baselines/floco/README.md:
```diff
@@ -0,0 +1,112 @@
+title: Federated Learning over Connected Modes
+url: https://openreview.net/pdf?id=JL2eMCfDW8
+labels: [personalization, cross-silo, linear mode connectivity] 
+dataset: [CIFAR10]
```
```suggestion
dataset: [CIFAR-10]
```

jafermarq reviewed (approved):
@dennis-grinwald, it's great to have `Floco` as part of the Flower Baselines! Thanks for the making it happen 🙌

dennis-grinwald commented:
Thanks!
